server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeBucketStrategy.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.composite;
+
+import org.elasticsearch.common.Rounding;
+import org.elasticsearch.common.geo.GeoBoundingBox;
+
+/**
+ * This class acts as a bit of syntactic sugar to let us pass in the rounding info for dates or the interval for numeric histograms as one
+ * class, to save needing three different interfaces.  Sometimes I miss C-style Union structures.
+ */
+public class CompositeBucketStrategy {
```

Comment:
It seems to me that this should be an interface that returns a `T extends ValuesSource` given a `ValuesSource`. 

examples

```
    interface StrategyApplier<T extends ValuesSource, R extends ValuesSource> {

        R build(T valueSource);

    }

    //Terms one
    public static class None implements StrategyApplier<ValuesSource, ValuesSource> {
        @Override
        public ValuesSource build(ValuesSource valuesSource) {
            return valuesSource;
        }
    }

    //Geotile one
    public static class Geo implements StrategyApplier<ValuesSource.GeoPoint, CellIdSource> {

        private final int precision;
        private final GeoBoundingBox boundingBox;
        
        public Geo(int precision, GeoBoundingBox boundingBox) {
            this.precision = precision;
            this.boundingBox = boundingBox;
        }

        @Override
        public CellIdSource build(ValuesSource.GeoPoint valueSource) {
            return new CellIdSource(
                valueSource,
                precision,
                boundingBox,
                GeoTileUtils::longEncode
            );
        }
    }

```

These appliers could constructed in your `innerBuild` objects. You will have to cast things, but it is doing that anyways in the `register` methods.

This seems cleaner to me than asserting on a strategy enum and having this pseudo union type.

Reply:
The problem is that we can't say that Geo always takes a `ValuesSource.GeoPoint`.  The whole goal here is to allow plugging in values sources from plugins without needing to touch the main aggregation code.  If we hard cast from `innerBuilder`, that breaks.  The hard casts need to happen in the registered `CompositeSupplier`s.  What I can do, and what we did for the other aggs, is instead of having one `CompositeSupplier` which takes this union thing, I can have four `CompositeSupplier` classes, one for each agg type, which take the appropriate rounding information.  I'd avoided doing that because it caused a lot of confusion when I did it last time, but maybe it's better than this?  IDK.

I'll code it up and we can compare.

Reply:
@not-napoleon 

> The problem is that we can't say that Geo always takes a ValuesSource.GeoPoint

Ok, maybe I misunderstand, but the geo composite strategy requires a `CellIdSource`, which requires a `ValuesSource.GeoPoint`. 

If you are adding flexibility somewhere, I don't see it.

Reply:
`CellIdSource` is just a way of turning a GeoPoint into a Long.  What `GeoTileValuesSource` requires is a function that yields longs, for which we currently pass in `CellIdSource#longValues()`.  Before this change, that was the only option, and was hard-coded in `CompositeAggregator`.  Essentially, we assumed that if you asked for a GeoTile composite, you must be using a geo point.  

The flexibility comes from being able to look up that logic at run time, rather than hard code it at compile time.  So, in the Spatial plugin, we have a class `GeoShapeCellSourceId`.  This change allows us to add a lookup that says "if you have a GeoShape, use the GeoShape version of CellSourceId instead".  That's done via that register method, which maps a values source type to a supplier function.  We can't put that logic directly into `innerBuilder` without adding a compile time dependency on the module, which we want to avoid.

Although I haven't written an example in here (in the interest of time and smaller change sets), the idea is that plugins which define new `ValuesSourceType`s, like Spatial defines GeoShape as a values source type, would register those types along with logic for how to turn those into what the key generation functions want.  In this way of thinking, it's "safe" to cast a values source from within a registered supplier function, because we know what type of the mapping key for that function, which indicates the type of values source we're getting.  It's not safe to cast from `innerBuilder`, because at that point we don't know what types we have, since a plugin could have added types at run time.  Make sense?

It's a little tricky to follow for composite because composite uses ValuesSource to mean three different things.  I'm going to do some follow up work to clean up some of that terminology.

Reply:
I wonder if it'd make sense to move the have three methods on `CompositeSupplier` that map to these three ctors. Each supplier could implement the strategy it supports and throw an exception for the ones it doesn't?